File: singly_ll.c
#include <stdio.h>
#include <stdlib.h>

// structures
typedef struct node node;

struct node {
    int data;
    node *next;
};

// prototypes / helper function declarations (main program)
node *createLL_rec(node *head, int data); // create a linked list recursively

node *createLL_insertByHead(node *head, int data); // create a LL by inserting a new node via head - LIFO
node *createLL_insertByTail(node *head, int data); // create a LL by inserting a new node via tail - FIFO

void printLL(node *head); // print out our LL

// functions that tackle interesting questions
node *insert_y_after_x(node *head, int x, int y);

int main(int argc, char const *argv[])
{
    int amount, value, i;
    node *ll = NULL;

    printf("How many nodes would you like to create? : ");
    scanf("%d", &amount);

    for (i = 0; i < amount; i++) {
        printf("Enter a value: ");
        scanf("%d", &value);
        //ll = createLL_insertByHead(ll, value);
        ll = createLL_insertByTail(ll, value);
    }
    printLL(ll);

    ll = insert_y_after_x(ll, 5, 99);

    printLL(ll);
    
    return 0;
}

node* createLL_insertByHead(node *head, int data) {
    node *temp = (node*)malloc(sizeof(node));
    temp->data = data;
    temp->next = NULL;
    node *hold = head;

    if (head == NULL) return temp;

    temp->next = hold;

    return temp;
}

node *createLL_insertByTail(node *head, int data) {
    node *temp = (node*)malloc(sizeof(node));
    temp->data = data;
    temp->next = NULL;
    node *iter = head;

    if (head == NULL) return temp;

    while (iter->next != NULL)
        iter = iter->next;

    iter->next = temp;
    
    return head;
}

void printLL(node *head) {
    node *temp = head;

    printf("Singly LL:\n");
    while (temp != NULL) {
        printf("%d%s", temp->data, temp->next != NULL ? " -> " : " -> NULL\n");
        temp = temp->next;
    }
}

node *insert_y_after_x(node *head, int x, int y) {
    node *iter = head;
    node *hold;
    
    while (iter != NULL) {
        if (iter->data == x) {
            hold = iter;
            node *temp = (node*)malloc(sizeof(node));
            temp->data = y;
            temp->next = hold->next;
            hold->next = temp;
        }
        iter = iter->next;
    }

    return head;
}

File: singly_ll.java
public class singly_ll {
    public Node head;

    public static void main(String[] args) {
        singly_ll objHead = new singly_ll();

        objHead.head = objHead.createLL_insertByTail(objHead.head, 5);
        objHead.head = objHead.createLL_insertByTail(objHead.head, 5);
        objHead.head = objHead.createLL_insertByTail(objHead.head, 43);
        objHead.head = objHead.createLL_insertByTail(objHead.head, 97);
        objHead.head = objHead.createLL_insertByTail(objHead.head, 5);
        objHead.head = objHead.createLL_insertByTail(objHead.head, 2);

        objHead.printLL(objHead.head);

        objHead.head = objHead.insert_y_after_x(objHead.head, 2, 81);

        objHead.printLL(objHead.head);
    }

    Node createLL_insertByTail(Node head, int data) {
        Node temp = new Node(null, data); // create new node
        Node iter = head;

        if (head == null) return temp;

        while (iter.next != null)
            iter = iter.next;

        iter.next = temp;

        return head; // we are returning the very first node which is the head so we are retaining this position
    }

    Node createLL_insertByHead(Node head, int data) {
        Node temp = new Node(null, data); // create new node

        if (head == null) return temp;

        temp.next = head;

        return temp; // we are returning this as the new head
    }

    /* insert a new node with the value y after a node with x value; so if y = 100 and x = 8, 
    we insert a new node with value 100 after a node with value 8. */
    Node insert_y_after_x(Node head, int x, int y) {
        Node iter = head, hold;

        while (iter != null) {
            if (iter.data == x) {
                hold = iter;
                Node temp = new Node(hold.next, y);
                hold.next = temp;
            }
            iter = iter.next;
        }

        return head;
    }

    // insert data in ascending order meaning that the head will have the lowest value and the tail will have the greatest value
    Node insert_inOrder(Node head, int data) {
        Node temp = new Node(null, data);
        Node iter = head, hold;

        if (head == null) return temp;

        if (head.data > data) { // if the head is greater than our current value
            temp.next = iter;
            return temp;
        }

        // we want to hold the node that less than our new node but the node ahead is greater than our new node as well, also checking for NULLs 
        while (iter.next.data < data && iter.next != null)
            iter = iter.next;

        hold = iter.next;
        iter.next = temp;
        temp.next = hold;

        return head;
    }

    void printLL(Node head) {
        Node iter = head;
        System.out.print("The Linked List:\n");
        while (iter != null) {
            System.out.printf("%d%s", iter.data, iter.next != null ? " -> " : " -> NULL\n");
            iter = iter.next;
        }
    }
}

class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }

    public Node(Node next, int data) {
        this.data = data;
        this.next = next;
    }
}